### DISCUSSION TITLE
How do you maintain userdata / AttributesTable when performing an operation?

### DISCUSSION THREAD

**Discussion by Kennethtruyers** [NetTopologySuite/NetTopologySuite#565] [Q&A]
I'm doing geo spatial work with NetTopoloySuite and C#. In Python (GeoPandas), you can do overlay operations on features (as opposed to on the raw geometries), so you can keep the metadata associated to those geometries.

Example (Python / GeoPandas):
```
features = geopandas.read_file("file_name")
otherFeatures = geopandas.read_file("otherFile")

overlay = geopandas.overlay(features , otherFeatures , how="intersection")
// overlay now contains the features that are in both data sets, together with the attributes
```

In NetTopologySuite, all operations are executed just on the Geometries:

```
var features = ReadShapeFile("file_name");
var otherFeatures = ReadShapeFile("otherFile");

// Overlay is only possible on geometries, not on features, so we have to extract the geometries
var geometries = new GeometryCollection(features.Select(f => f.Geometry).ToArray());
var otherGeometries = new GeometryCollection(otherFeatures.Select(f => f.Geometry).ToArray());

var overlay = new OverlayNG(geometries, otherGeometries, SpatialFunction.Intersection);
// Overlay now contains the geometries that are in both data sets, but it doesn't contain the original attributes
```

Is there a way to do spatial operations on Features in NetTopologySuite that retains the metadata associated to those features?

(For completeness sake, the code behind ReadShapeFile)

```
public static IEnumerable<Feature> ReadShapeFile(string path)
{
    using (var rd = new ShapefileDataReader(path, GeometryFactory.Default))
    {
        var fieldNames = Enumerable.Range(1, rd.DbaseHeader.NumFields)
                                   .Select(i => rd.GetName(i))
                                   .ToList();

        while (rd.Read())
            yield return new Feature(rd.Geometry, new AttributesTable(fieldNames.Select((n, i) => new KeyValuePair<string, object>(n, rd.GetValue(i + 1)))));
        }
}
```

**Comment by FObermaier** [comment]
There is no such functionality in these projects. This would make a nice addition to `NetTopologySuite.Features` project.

**Reply by Kennethtruyers** [reply]
Do you have any pointers on how it could be accomplished? If I can find a way of doing this, I could raise a PR for that project

**Reply by Kennethtruyers** [reply]
I've been looking a bit at this and I don't think it can be just added to the NetTopologysuite.Features project. When you do an operation such as Overlay, NTS creates new Geometries. In order to carry the attributes, I'd have to modify the internals of those operations so they attach the attributes to the geometries and carry them through every manipulation. Am I missing something?

**Reply by DGuidi** [reply]
> // overlay now contains the features that are in both data sets, together with the attributes

if you talk about selecting features of two datasets that matches some condition, you can maybe use the UserData field to store the Id of any feature? 
for union/intersect operation usually you have new geometries, so I don't know how geopandas decide the attributes to copy to the resulting geometry. do you know how to retrieve some docs about the inner logic of those operations?

**Reply by Kennethtruyers** [reply]
I've had a quick look at the source code of GeoPandas, more specifically the overlay. When it does an intersecting overlay, it essentially creates pairs of geometries from both datasets (so n*m pairs) and intersects each all of them. When intersecting a pair, it merges the metadata from each of those geometries into the resulting intersection. It then filters out all the ones that have an actual intersection and returns that result. 

When doing a difference overlay, it just keeps the original metadata. All other operations are just a combination of those two operations. 

There are no real docs on those operations, but you can see it directly in the source code:
https://github.com/geopandas/geopandas/blob/master/geopandas/tools/overlay.py#L32

I suppose I could work out something like that, but it would mean either duplicating all the functionality in the features library or a fundamental rewrite of the way it's implemented inside NTS (both of which are a bit too big for me to tackle)

**Reply by DGuidi** [reply]
> it merges the metadata from each of those geometries 
for "metadata" do you mean the attributes?

> duplicating all the functionality in the features library or a fundamental rewrite of the way it's implemented inside NTS
maybe a library on TOP of NTS? from your description, the userdata field in any geometry can be enough, but actually I'm not sure how it is "persisted" across the operations... 

In any case, looks an interesting feature and a not-so-small effort.

**Reply by Kennethtruyers** [reply]
Yeah, the userdata field would be enough as that way we could potentially update the Features library to use the UserData object as a backing field for the attributes table, rather than a separate object. 

I'm not sure how you could implement it on top of NTS though, since the new geometries are created inside the operations. If we'd implement it on top of NTS, we'd have to run a separate operation to then go through the created geometries. 

I found it weird that this is not implemented though, it seems like a basic requirement for anyone working with geospatial data. Geometries usually don't exist in a void, they tend to have data associated with them. 

I'll have a play with just implementing the overlay on the features library, similar to what GeoPandas is doing

**Reply by DGuidi** [reply]
NTS is a port of JTS, so mainly topology and geometry operations. there's a lot of space for other libraries to fill the gaps :)

**Reply by Kennethtruyers** [reply]
Made a quick proof of concept PR for these operations here: https://github.com/NetTopologySuite/NetTopologySuite.Features/pull/14/files

Mind having a look and let me know what you think?

**Reply by DGuidi** [reply]
I will do ASAP, thanks :)
